src/adapter/src/coord/sequencer/inner/create_materialized_view.rs

```diff
@@ -540,8 +540,69 @@ impl Coordinator {
     ) -> Result<StageResult<Box<CreateMaterializedViewStage>>, AdapterError> {
         // Timestamp selection
         let id_bundle = dataflow_import_id_bundle(global_lir_plan.df_desc(), cluster_id);
-        let (read_holds, dataflow_as_of, storage_as_of, until) =
-            self.select_timestamps(id_bundle, refresh_schedule.as_ref())?;
+
+        let read_holds_owned;
+        let read_holds = if let Some(txn_reads) = self.txn_read_holds.get(session.conn_id()) {
+            // In some cases, for example when REFRESH is used, the preparatory
+            // stages will already have acquired ReadHolds, we can re-use those.
+            tracing::info!(session = %session.conn_id(), ?txn_reads, "using txn read holds for materialized view");
+
+            // Verify that all acquired read holds are for exactly the same
+            // bundle of IDs, which are also the reads that we have determined
+            // above.
+            for read_holds in txn_reads.iter() {
+                assert!(
```

Comment:
This seems a bit brittle. Both `resolve_mz_now_for_create_materialized_view` and `create_materialized_view_validate` call `sufficient_collections`, but can we be sure that no index is dropped or created in-between? (`sufficient_collections` includes all indexes on all the referenced input collections.)

Reply:
would it be more correct to assert that all the read holds have _at least_ the collections in the `id_bundle` that we determine right here in the method?

Reply:
I hate to say it, but it might be time to add multiplicities to `ReadHolds` so that we can get rid of `Vec<ReadHolds>` and replace it with a single `ReadHolds` per transaction. It feels like this block is making a lot of assumptions about the `txn_reads` that are not enforced anywhere and may be easy to accidentally break in the future.

Another option is maybe we could make a temporary `ReadHolds` object here by combining all of the `ReadHolds` in the `Vec` and use that for selecting the timestamp.

Reply:
You might be right! let me sit on that over the weekend